src/big/big_decimal.cr

```diff
@@ -47,34 +47,79 @@ struct BigDecimal < Number
   def initialize(str : String)
     raise InvalidBigDecimalException.new(str, "Zero size") if str.bytesize == 0
 
-    # Check str's validity and find index of .
+    # Check str's validity and find index of '.'
     decimal_index = nil
+    # Check str's validity and find index of 'e'
+    exponent_index = nil
+
     str.each_char_with_index do |char, index|
       case char
       when '-'
-        if index != 0
+        unless index == 0 || exponent_index == index - 1
           raise InvalidBigDecimalException.new(str, "Unexpected '-' character")
         end
+      when '+'
+        unless exponent_index == index - 1
+          raise InvalidBigDecimalException.new(str, "Unexpected '+' character")
+        end
       when '.'
         if decimal_index
           raise InvalidBigDecimalException.new(str, "Unexpected '.' character")
         end
-
         decimal_index = index
+      when 'e', 'E'
+        if exponent_index
+          raise InvalidBigDecimalException.new(str, "Unexpected #{char.inspect} character")
+        end
+        exponent_index = index
       when '0'..'9'
         # Pass
       else
         raise InvalidBigDecimalException.new(str, "Unexpected #{char.inspect} character")
       end
     end
 
-    if decimal_index
+    case
+    when exponent_index
+      exponent_postfix = str[exponent_index + 1]
+      case exponent_postfix
+      when '+', '-'
+        exponent_positive = exponent_postfix == '+'
+        exponent = str[exponent_index + 2..-1].to_u64
+      else
+        exponent_positive = true
+        exponent = str[exponent_index + 1..-1].to_u64
+      end
+      if decimal_index
+        decimals = (exponent_index - decimal_index - 1).to_u64
```

Comment:
By defining
```
decimals = ((exponent_index || str.bytesize) - decimal_index - 1)
```

I think we can remove a lot of duplication here. We don't need to duplicate the `@value` calculation for each branch, just have
```
if decimal_index
  # the existing code but with the above calculation
end

if exponent_index
  # ...
end
```

Reply:
I'm not sure if I follow... to define

```cr
decimals = ((exponent_index || str.bytesize) - decimal_index - 1)
```

you'd need to do it inside `if decimal_index` branch, otherwise compiler will complain about possible `nil` in `decimal_index - 1`. that's where I'm lost... perhaps some code could to illustrate your point could be a help here.

Reply:
I absolutely haven't tested this:

```cr
decimal_end_index = (exponent_index || str.bytesize) - 1
if decimal_index # rename to decimal_start_index ?
  value_str = String.build do |builder|
    # We know this is ASCII, so we can slice by index
    builder.write(str.to_slice[0, decimal_index])
    builder.write(str.to_slice[decimal_index + 1, decimal_end_index - decimal_index])
  end
  @value = value_str.to_big_i
else
  @value = str[0..decimal_end_index].to_big_i
end

if decimal_index
  decimal_count = decimal_end_index - decimal_index
else
  decimal_count = 0
end

if exponent_index
  exponent_postfix = str[exponent_index + 1]
  case exponent_postfix
  when '+', '-'
    exponent_positive = exponent_postfix == '+'
    exponent = str[(exponent_index + 2)..-1].to_u64
  else
    exponent_positive = true
    exponent = str[(exponent_index + 1)..-1].to_u64
  end

  @scale = exponent
  if exponent_positive
    @scale -= decimal_count
    @value *= 10.to_big_i ** @scale
    @scale = 0_u64
  else
    @scale += decimal_count
  end
else
  @scale = decimal_count.to_u64
end
```

Reply:
I don't entirely understand why you do
```cr
    @value *= 10.to_big_i ** @scale
    @scale = 0_u64
```

in your code, couldn't you just take these lines out or am I being stupid?